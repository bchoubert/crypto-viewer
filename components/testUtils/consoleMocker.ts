/* eslint-disable no-console */
const consoleMocker = {
  consoleError: null,

  errorsToSuppress: [
    'Failed prop type',
  ],

  mockConsole: () => {
    consoleMocker.consoleError = console.error;
    console.error = (message: string) => {
      let printMessage = true;

      consoleMocker.errorsToSuppress.forEach((errorToSuppress: string) => {
        if (message.includes(errorToSuppress)) {
          printMessage = false;
        }
      });

      if (printMessage) {
        consoleMocker.consoleError?.(message);
      }
    };
  },

  restoreConsole: () => {
    if (consoleMocker.consoleError) {
      console.error = consoleMocker.consoleError;
    }
  },
};

export default consoleMocker;
