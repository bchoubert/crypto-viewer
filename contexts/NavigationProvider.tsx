import React, {
  createContext, FC, ReactNode, useCallback, useState, useEffect, useMemo, memo, useContext,
} from 'react';
import { StatusBarStyle, BackHandler } from 'react-native';
import { Colors } from 'react-native/Libraries/NewAppScreen';

import Tabs, { TabType } from '../models/Tabs';
import Crypto from '../models/Crypto';
import ColorService from '../services/Color.service';
import { ThemeContext } from './ThemeProvider';

export const NavigationContext = createContext({
  activeTab: Tabs.list as TabType,
  details: null as Crypto | null,
  statusBarColor: Colors.white as string,
  statusBarMode: 'dark-content' as StatusBarStyle,
  changeTab: null as (tabName: TabType, newDetails?: Object) => void | null,
});

interface NavigationProviderProps {
  children: ReactNode;
}

const NavigationProvider: FC<NavigationProviderProps> = ({ children }) => {
  const theme = useContext(ThemeContext);

  const [activeTab, setActiveTab] = useState<TabType>(Tabs.list);

  // Store details about current active tab
  const [details, setDetails] = useState<Crypto | null>(null);

  // StatusBar mode
  const [statusBarColor, setStatusBarColor] = useState<string>(theme.backgroundColor);
  const [statusBarMode, setStatusBarMode] = useState<StatusBarStyle>('dark-content');

  const changeTab = useCallback((tabName: TabType, newDetails: Crypto = null) => {
    setStatusBarColor(tabName === Tabs.details
      ? ColorService.getColorFromCrypto(newDetails.id)
      : theme.backgroundColor);
    setStatusBarMode(tabName === Tabs.details ? 'light-content' as StatusBarStyle : 'dark-content' as StatusBarStyle);
    setDetails(newDetails);
    setActiveTab(tabName);
  }, []);

  const handleBackButton = useCallback(() => {
    if ([Tabs.settings, Tabs.details].includes(activeTab)) {
      changeTab(Tabs.list);
      return true;
    }
    return false;
  }, [activeTab, changeTab]);

  useEffect(() => {
    const backHandler = BackHandler.addEventListener('hardwareBackPress', handleBackButton);
    return () => backHandler.remove();
  }, [activeTab]);

  const contextValues = useMemo(() => ({
    activeTab,
    details,
    statusBarColor,
    statusBarMode,
    changeTab,
  }), [activeTab, details, statusBarColor, statusBarMode, changeTab]);

  useEffect(() => {
    setStatusBarColor(theme.backgroundColor);
  }, [theme]);

  return (
    <NavigationContext.Provider value={contextValues}>
      {children}
    </NavigationContext.Provider>
  );
};

export default memo(NavigationProvider);
