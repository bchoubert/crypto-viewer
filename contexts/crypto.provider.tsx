import { showError } from "@/services/error.service";
import { fetchCryptoExchangeRates, fetchCryptos } from "@/services/network.service";
import { ICrypto } from "@/types/crypto.types";
import { FC, ReactNode, createContext, memo, useCallback, useContext, useEffect, useMemo, useState } from "react";
import { ToastAndroid } from "react-native";
import { SettingsContext } from "./settings.provider";
import { TranslationsContext } from "./translations.provider";

interface CryptoContextInterface {
  cryptos: ICrypto[];
  isLoading: boolean;
  refreshCryptos: () => void;
}

export const CryptoContext = createContext<CryptoContextInterface>({
  cryptos: [],
  isLoading: false,
  refreshCryptos: () => {},
});

interface CryptoProviderProps {
  children: ReactNode;
}

const CryptoProvider: FC<CryptoProviderProps> = memo(({
  children
}) => {
  const { settings } = useContext(SettingsContext);
  const translation = useContext(TranslationsContext);

  const [isLoading, setLoading] = useState<boolean>(true);
  const [cryptos, setCryptos] = useState<ICrypto[]>([]);

  const refreshCryptos = useCallback(async () => {
    setLoading(true);

    const [fetchedCurrencies, exchangeRatesResponse] = await Promise.all([
      fetchCryptos().catch(showError),
      fetchCryptoExchangeRates(settings.quote).catch(showError),
    ]);

    if (!fetchedCurrencies) {
      ToastAndroid.show(translation.list.error, ToastAndroid.BOTTOM);
      return;
    }
    
    const rates = (exchangeRatesResponse?.data?.rates || {});
    const fetchedCurrenciesWithRates: ICrypto[] = fetchedCurrencies.filter(c => c?.details?.type === 'crypto').map(f => ({
      ...f,
      rate: rates[f.id] ? (1 / parseFloat(rates[f.id])) : undefined,
    }));

    const settingSortAttribute = settings.sortAssets;

    const sortedCurrenciesWithRates = fetchedCurrenciesWithRates.sort((a, b) => {
      const valueA = a[settingSortAttribute].toLowerCase();
      const valueB = b[settingSortAttribute].toLowerCase();

      if (valueA < valueB) {
        return -1;
      }
      if (valueA > valueB) {
        return 1;
      }
      return 0;
    });

    setCryptos(sortedCurrenciesWithRates);
    setLoading(false);
  }, [settings]);

  useEffect(() => { refreshCryptos() }, [settings]);

  const contextValues: CryptoContextInterface = useMemo(() => ({
    cryptos,
    isLoading,
    refreshCryptos,
  }), [cryptos, refreshCryptos, isLoading]);

  return (
    <CryptoContext.Provider value={contextValues}>
      {children}
    </CryptoContext.Provider>
  );
})

export default CryptoProvider;
