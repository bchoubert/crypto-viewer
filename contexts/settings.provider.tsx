import { AvailableTranslations } from "@/assets/translations/TranslationUtils";
import { loadSettings, saveSettings } from "@/services/settings.service";
import { Quote, SortAssetsType } from "@/types/crypto.types";
import { DarkModeType } from "@/types/darkMode.types";
import { DateFormatType } from "@/types/date.types";
import { GraphModeType } from "@/types/graph.types";
import { SettingsEnum, SettingsType, defaultSettings, settingsDetails } from "@/types/settings.types";
import { TranslationPossibility, TranslationType } from "@/types/translation.types";
import { WalletItem } from "@/types/wallet.types";
import { FC, ReactNode, createContext, memo, useCallback, useEffect, useMemo, useState } from "react";

interface SettingsContextInterface {
  settings: SettingsType;
  favourites: string[];
  wallet: WalletItem[];

  changeSetting: (settingKey: SettingsEnum, value: string, translation: TranslationType) => void;
  
  toggleFavourite: (id: string) => void;
  hasFavourite: (id: string) => boolean;

  addWalletItem: (item: WalletItem) => void;
  removeWalletItem: (id: string) => void;
}

export const SettingsContext = createContext({
  settings: defaultSettings as SettingsType,
} as SettingsContextInterface);

interface SettingsProviderProps {
  children: ReactNode;
}

const SettingsProvider: FC<SettingsProviderProps> = memo(({
  children,
}) => {
  const [isLoading, setLoading] = useState<boolean>(true);
  const [values, setValues] = useState<SettingsType>(defaultSettings);

  const load = useCallback(async () => {
    const settings = await loadSettings();
    setValues(settings);
    setLoading(false);
  }, []);

  useEffect(() => { load() }, []);

  const changeSettingWithValue = useCallback((settingKey: keyof SettingsType, value: any) => {
    const newValues = {
      ...values,
      [settingKey]: value
    };

    setValues(newValues);
    saveSettings(newValues);
  }, [values, setValues]);

  const changeSettingViaAccessor = useCallback((settingKey: SettingsEnum, value: any, translation: TranslationType) => {
    const accessor = settingsDetails[settingKey](translation).accessor;

    changeSettingWithValue(accessor, value);
  }, [changeSettingWithValue]);
  
  const hasFavourite = useCallback((id: string) => values.favourites.includes(id), [values]);
  const toggleFavourite = useCallback((id: string) => {
    const isNotFavourite = !hasFavourite(id);
    const newFav = [...values.favourites].filter(i => i !== id);

    if (isNotFavourite) {
      changeSettingWithValue('favourites', [...newFav, id]);
    } else {
      changeSettingWithValue('favourites', newFav);
    }
  }, [changeSettingWithValue, values]);


  const addWalletItem = useCallback((item: WalletItem) => {
    const newWallet = [...values.wallet].filter(w => w.id !== item.id);
    changeSettingWithValue('wallet', [...newWallet, item]);
  }, [changeSettingWithValue, values]);

  const removeWalletItem = useCallback((id: string) => {
    const newWallet = [...values.wallet].filter(w => w.id !== id);
    changeSettingWithValue('wallet', newWallet);
  }, [changeSettingWithValue, values]);

  const contextValues = useMemo(
    () => {
      const context: SettingsContextInterface = {
        settings: values,
        favourites: values.favourites,
        wallet: values.wallet,

        changeSetting: changeSettingViaAccessor,

        toggleFavourite,
        hasFavourite,

        addWalletItem,
        removeWalletItem,
      };
      
      return context;
    },
    [ values, changeSettingViaAccessor, toggleFavourite, hasFavourite, addWalletItem, removeWalletItem],
  );

  if (isLoading) {
    return null;
  }

  return (
    <SettingsContext.Provider value={contextValues}>
      {children}
    </SettingsContext.Provider>
  )
});

export default SettingsProvider;
