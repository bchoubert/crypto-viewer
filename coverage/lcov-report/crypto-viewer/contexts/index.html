
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for crypto-viewer/contexts</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> crypto-viewer/contexts</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">50.63% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>40/79</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">0% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>0/36</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">52.38% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>11/21</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">49.32% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>36/73</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
    </div>
    <div class='status-line medium'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file medium" data-value="NavigationProvider.tsx"><a href="NavigationProvider.tsx.html">NavigationProvider.tsx</a></td>
	<td data-value="65.52" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 65%"></div><div class="cover-empty" style="width: 35%"></div></div>
	</td>
	<td data-value="65.52" class="pct medium">65.52%</td>
	<td data-value="29" class="abs medium">19/29</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="11" class="abs low">0/11</td>
	<td data-value="71.43" class="pct medium">71.43%</td>
	<td data-value="7" class="abs medium">5/7</td>
	<td data-value="62.96" class="pct medium">62.96%</td>
	<td data-value="27" class="abs medium">17/27</td>
	</tr>

<tr>
	<td class="file low" data-value="SettingsProvider.tsx"><a href="SettingsProvider.tsx.html">SettingsProvider.tsx</a></td>
	<td data-value="10.53" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 10%"></div><div class="cover-empty" style="width: 90%"></div></div>
	</td>
	<td data-value="10.53" class="pct low">10.53%</td>
	<td data-value="19" class="abs low">2/19</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="6" class="abs low">0/6</td>
	<td data-value="11.76" class="pct low">11.76%</td>
	<td data-value="17" class="abs low">2/17</td>
	</tr>

<tr>
	<td class="file low" data-value="ThemeProvider.tsx"><a href="ThemeProvider.tsx.html">ThemeProvider.tsx</a></td>
	<td data-value="47.83" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 47%"></div><div class="cover-empty" style="width: 53%"></div></div>
	</td>
	<td data-value="47.83" class="pct low">47.83%</td>
	<td data-value="23" class="abs low">11/23</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="23" class="abs low">0/23</td>
	<td data-value="60" class="pct medium">60%</td>
	<td data-value="5" class="abs medium">3/5</td>
	<td data-value="45.45" class="pct low">45.45%</td>
	<td data-value="22" class="abs low">10/22</td>
	</tr>

<tr>
	<td class="file high" data-value="TranslationProvider.tsx"><a href="TranslationProvider.tsx.html">TranslationProvider.tsx</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank">istanbul</a>
                at Mon Sep 06 2021 22:37:01 GMT+0200 (Central European Summer Time)
            </div>
        </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    