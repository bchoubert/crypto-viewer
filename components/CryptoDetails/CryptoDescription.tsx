import React, {
  FC, useCallback, useMemo, useContext,
} from 'react';
import {
  View, StyleSheet, Text, TouchableOpacity, Linking,
} from 'react-native';

import UtilsService from '../../services/Utils.service';

import Colors from '../../assets/Colors';

import CryptoViewerIconsMap from '../../assets/fonts/baseIcons/CryptoViewerIconsMap';
import CryptoCurrenciesIconMap from '../Utils/CryptoCurrencyIconsMap';

import { NavigationContext } from '../../contexts/NavigationProvider';

const styles = StyleSheet.create({
  description: {
    flexDirection: 'column',
  },
  description_title: {
    paddingLeft: 20,
    paddingRight: 10,
    fontSize: 20,
    fontWeight: 'bold',
  },
  description_text: {
    color: Colors.darkGray,
    paddingHorizontal: 20,
    paddingVertical: 10,
    fontSize: 15,
    lineHeight: 20,
  },
  description_website: {
    paddingHorizontal: 20,
    paddingBottom: 10,
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
  },
  description_website_icon: {
    fontSize: 12,
  },
  description_website_text: {
    fontWeight: 'bold',
    width: 95,
  },

  cryptoViewerIcon: {
    fontSize: 20,
    fontFamily: 'crypto-viewer',
  },
});

interface CryptoDescriptionProps {}

const CryptoDescription: FC<CryptoDescriptionProps> = () => {
  const {
    details,
  } = useContext(NavigationContext);

  const cryptoColor = useMemo(
    () => UtilsService.getColorFromCrypto(details.id),
    [details],
  );

  const openWebsite = useCallback(
    () => Linking.openURL(CryptoCurrenciesIconMap[details.id?.toLowerCase()]?.website),
    [details],
  );

  return (CryptoCurrenciesIconMap[details.id?.toLowerCase()]?.description
    || CryptoCurrenciesIconMap[details.id?.toLowerCase()]?.website ? (
      <View style={styles.description}>
        <Text style={styles.description_title}>
          Details
        </Text>
        <Text style={styles.description_text}>
          {CryptoCurrenciesIconMap[details.id?.toLowerCase()]?.description || ''}
        </Text>
        {CryptoCurrenciesIconMap[details.id?.toLowerCase()]?.website ? (
          <TouchableOpacity onPress={openWebsite} style={styles.description_website}>
            <Text
              numberOfLines={1}
              style={{ ...styles.description_website_text, color: cryptoColor }}
            >
              Official Website
            </Text>
            <Text
              style={{
                ...styles.cryptoViewerIcon,
                ...styles.description_website_icon,
                color: cryptoColor,
              }}
            >
              {CryptoViewerIconsMap.link.unicode}
            </Text>
          </TouchableOpacity>
        ) : null}
      </View>
    ) : null
  );
};

export default CryptoDescription;
