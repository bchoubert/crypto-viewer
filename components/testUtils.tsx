/* eslint-disable import/no-extraneous-dependencies */
import { render, RenderAPI } from '@testing-library/react-native';
import '@testing-library/jest-dom/extend-expect';
import React, { ReactElement } from 'react';

import NavigationProvider from '../contexts/NavigationProvider';
import { SettingsContext } from '../contexts/SettingsProvider';
import ThemeProvider from '../contexts/ThemeProvider';
import TranslationProvider from '../contexts/TranslationProvider';

import { defaultSettings } from '../models/SettingsType';

jest.useFakeTimers();

export const enrichProviders = (node: ReactElement): ReactElement => (
  <SettingsContext.Provider
    value={{
      settings: defaultSettings,
      changeSettings: jest.fn(),
    }}
  >
    <TranslationProvider>
      <ThemeProvider>
        <NavigationProvider>
          {node}
        </NavigationProvider>
      </ThemeProvider>
    </TranslationProvider>
  </SettingsContext.Provider>
);

// eslint-disable-next-line max-len
export const renderNode = (node: ReactElement): RenderAPI => render(enrichProviders(node));
