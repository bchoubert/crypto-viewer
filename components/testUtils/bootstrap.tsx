/* eslint-disable no-console */
/* eslint-disable import/no-extraneous-dependencies */
import { render, RenderAPI } from '@testing-library/react-native';
import '@testing-library/jest-dom/extend-expect';
import React, { ReactElement } from 'react';
import * as PassThoughComponents from '../PassThroughComponents';

import NavigationProvider from '../../contexts/NavigationProvider';
import { SettingsContext } from '../../contexts/SettingsProvider';
import ThemeProvider from '../../contexts/ThemeProvider';
import TranslationProvider from '../../contexts/TranslationProvider';

import { defaultSettings } from '../../models/SettingsType';
import consoleMocker from './consoleMocker';

jest.useFakeTimers();

export const enrichProviders = (node: ReactElement): ReactElement => (
  <SettingsContext.Provider
    value={{
      settings: defaultSettings,
      changeSettings: jest.fn(),
    }}
  >
    <TranslationProvider>
      <ThemeProvider>
        <NavigationProvider>
          {node}
        </NavigationProvider>
      </ThemeProvider>
    </TranslationProvider>
  </SettingsContext.Provider>
);

// eslint-disable-next-line max-len
export const renderNode = (node: ReactElement): RenderAPI => render(enrichProviders(node));

beforeAll(() => {
  consoleMocker.mockConsole();
  // eslint-disable-next-line no-extend-native
  Date.prototype.getTimezoneOffset = jest.fn(() => 0);
});

afterAll(() => {
  consoleMocker.restoreConsole();
  jest.resetAllMocks();
});

jest.spyOn(PassThoughComponents, 'Image');
(PassThoughComponents.Image as unknown as jest.Mock).mockImplementation(() => (<div>IMAGE</div>));
