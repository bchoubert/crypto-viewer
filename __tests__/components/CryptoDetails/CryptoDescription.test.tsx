import React from 'react';
import CryptoDescription from '../../../components/CryptoDetails/CryptoDescription';
import { renderNode } from '../../../components/testUtils/bootstrap';
import CryptoCurrenciesIconMap from '../../../components/Utils/CryptoCurrencyIconsMap';
import { NavigationContext, NavigationContextInterface } from '../../../contexts/NavigationProvider';

describe('<CryptoDescription />', () => {
  const navigationContextValue: NavigationContextInterface = {
    statusBarColor: '#334455',
    statusBarMode: 'default',
    activeTab: 'details',
    changeTab: () => null,
    handleBackAction: () => null,
    details: {
      price: 2,
      id: 'btc',
      name: 'Bitcoin',
      status: 'online',
      details: {
        type: '',
        symbol: 'B',
      },
    },
  };

  const cryptoDetails = CryptoCurrenciesIconMap.btc;

  it('Prints correctly', () => {
    const { toJSON } = renderNode(
      <NavigationContext.Provider value={navigationContextValue}>
        <CryptoDescription />
      </NavigationContext.Provider>,
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('Basic behaviour', () => {
    const node = renderNode(
      <NavigationContext.Provider value={navigationContextValue}>
        <CryptoDescription />
      </NavigationContext.Provider>,
    );

    node.getByText(cryptoDetails.description);

    expect(node.getByTestId('website-text').props.accessibilityLabel).toContain(cryptoDetails.website);
  });
});
