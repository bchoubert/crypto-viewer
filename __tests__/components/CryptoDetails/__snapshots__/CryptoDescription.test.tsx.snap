// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CryptoDescription /> Prints correctly 1`] = `
<View
  style={
    Object {
      "flexDirection": "column",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#222222",
        "fontSize": 20,
        "fontWeight": "bold",
        "paddingLeft": 20,
        "paddingRight": 10,
      }
    }
  >
    Details
  </Text>
  <Text
    style={
      Object {
        "color": "#222222",
        "fontSize": 15,
        "lineHeight": 20,
        "paddingHorizontal": 20,
        "paddingVertical": 10,
      }
    }
  >
    The world’s first cryptocurrency, Bitcoin is stored and exchanged securely on the internet through a digital ledger known as a blockchain. Bitcoins are divisible into smaller units known as satoshis — each satoshi is worth 0.00000001 bitcoin.
  </Text>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "opacity": 1,
        "paddingBottom": 10,
        "paddingHorizontal": 20,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#F7931B",
          "fontFamily": "crypto-viewer",
          "fontSize": 12,
          "marginLeft": 5,
          "width": 10,
        }
      }
    >
      
    </Text>
    <Text
      accessibilityLabel="Go to https://bitcoin.org/"
      style={
        Object {
          "color": "#F7931B",
          "fontWeight": "bold",
          "marginLeft": 10,
          "width": 100,
        }
      }
      testID="website-text"
    >
      Official Website
    </Text>
  </View>
</View>
`;
